### PR TITLE
libraries: add notification settings

### PR DESCRIPTION
**Pull request by AoNoOokami**
* Improves library editor appearance for large screens. It doesn't take
the full width anymore.
* Adds tabs in the library detailed view.
* Adds notifications settings in library custom editor.
* Displays notification settings in library detail view.

Co-Authored-by: Alicia Zangger <alicia.zangger@rero.ch>
Co-Authored-by: Johnny Mariéthoz <johnny.mariethoz@rero.ch>
Co-Authored-by: Alicia Zangger <alicia.zangger@rero.ch>

## Why are you opening this PR?
To implement task 2013 of US1922 https://tree.taiga.io/project/rero21-reroils/us/1922?milestone=282105

## Dependencies

* https://github.com/rero/rero-ils/pull/1691

## How to test?

- What command should I have to run to test your PR?
- What should I test through the UI?

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: projects/admin/src/app/record/custom-editor/libraries/library.component.ts [hunk 1]
```diff
@@ -25,17 +25,23 @@ import { ToastrService } from 'ngx-toastr';
...
-export class LibraryComponent implements OnInit {
+export class LibraryComponent implements OnInit, OnDestroy {
 
   public library: Library;
   public libForm: FormGroup;
-  public organisationPid;
+  public organisationPid: string;
+
+  /** List of internal notification types */
+  internalTypes = ['booking', 'request', 'transit_notice'];
```
Comment (zannkukai) [inline projects/admin/src/app/record/custom-editor/libraries/library.component.ts:42]: Should be set in a Enum class into `class/library.ts` file. As ItemStatus, or NoteType

**Review by zannkukai — CHANGES_REQUESTED** [review]

File: projects/admin/src/app/classes/library.ts [hunk 2]
```diff
@@ -19,9 +19,14 @@
 // required as json properties is not lowerCamelCase
 
 import { WeekDay } from '@angular/common';
+import { marker } from '@biesbjerg/ngx-translate-extract-marker';
```
Comment (Garfield-fr) [inline projects/admin/src/app/classes/library.ts:22]: `import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';`

File: projects/admin/src/app/classes/library.ts [hunk 3]
```diff
@@ -19,9 +19,14 @@
...
 import * as moment from 'moment';
 import { WeekDays } from './week-days';
 
+export function _(str) {
+  return marker(str);
+}
+
```
Comment (Garfield-fr) [inline projects/admin/src/app/classes/library.ts:29]: Suggestion: Remove this (test extraction)

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 4]
```diff
@@ -29,14 +31,22 @@ export class LibraryFormService {
 
   public form;
 
+
+  private notificationTypes = [];
+
+  /** Observable for build event */
+  private buildEvent = new Subject();
+
   constructor(
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:40]: Missing doc string

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 5]
```diff
@@ -29,14 +31,22 @@ export class LibraryFormService {
 
   public form;
 
+
+  private notificationTypes = [];
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:35]: missing doc string

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 6]
```diff
@@ -29,14 +31,22 @@ export class LibraryFormService {
...
-      this.build();
-  }
+    private _fb: FormBuilder,
+    private _recordService: RecordService
+    ) { }
 
+  /**
+   * Build form
+   */
   build() {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:48]: `build(): void {`

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 7]
```diff
@@ -48,24 +58,51 @@ export class LibraryFormService {
...
+      notification_settings: this._fb.array([])
     });
     this.initializeOpeningHours();
+    this.initializeNotificationSettings();
+  }
+
+  /**
+   * Get build event
+   */
+  getBuildEvent() {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:71]: `getBuildEvent(): Observable<boolean> {`

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 8]
```diff
@@ -48,24 +58,51 @@ export class LibraryFormService {
...
+    return this.buildEvent.asObservable();
+  }
+
+  create() {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:75]: `create(): void {`

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 9]
```diff
@@ -48,24 +58,51 @@ export class LibraryFormService {
...
+        this.build();
+        this.buildEvent.next(true);
+      });
   }
 
+  /**
+   * Build and set default values for opening hours at form initialization
+   * @param openingHours - opening hours
+   */
   initializeOpeningHours(openingHours = []) {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:89]: `initializeOpeningHours(openingHours = []): void {`

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 10]
```diff
@@ -48,24 +58,51 @@ export class LibraryFormService {
...
       ));
     }
     this.setOpeningHours(openingHours);
   }
 
+  /**
+   * Set opening hours from record data
+   * @param openingHours - opening hours
+   */
   setOpeningHours(openingHours = []) {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:106]: `setOpeningHours(openingHours = []): void {`

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 11]
```diff
@@ -115,10 +163,10 @@ export class LibraryFormService {
...
 
-  reset() {
-    this.build();
-  }
-
+  /**
+   * Populate the form
+   * @param library - library
+   */
   populate(library: Library) {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:170]: `populate(library: Library): void {`

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 12]
```diff
@@ -127,6 +175,56 @@ export class LibraryFormService {
...
     });
     this.setOpeningHours(library.opening_hours);
+    this.setNotificationSettings(library.notification_settings);
+  }
+
+  /**
+   * Build an set default values for notification settings
+   * @param notificationSettings  - notification settings
+   */
+  initializeNotificationSettings(notificationSettings = []) {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:185]: Add return type

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 13]
```diff
@@ -127,6 +175,56 @@ export class LibraryFormService {
...
+      settings.push(this.getSettingsByType(type));
+    });
+    this.setNotificationSettings(notificationSettings);
+  }
+
+  /**
+   * Get setting by type
+   * @param settingType - setting type
+   */
+  getSettingsByType(settingType: NotificationType) {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:197]: Add return type

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 14]
```diff
@@ -127,6 +175,56 @@ export class LibraryFormService {
...
+        break;
+    }
+    return this._fb.group(model);
+  }
+
+  /**
+   * Set values from record
+   * @param notificationSettings - notification settings
+   */
+  setNotificationSettings(notificationSettings = []) {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:214]: Add return type

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 15]
```diff
@@ -140,6 +238,9 @@ export class LibraryFormService {
   get opening_hours() {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:238]: Add return type

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 16]
```diff
@@ -140,6 +238,9 @@ export class LibraryFormService {
...
     return this.form.get('opening_hours') as FormArray;
   }
+  get notification_settings() {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:241]: Missing doc string and return type

File: projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts [hunk 17]
```diff
@@ -140,6 +238,9 @@ export class LibraryFormService {
...
+    return this.form.get('notification_settings') as FormArray;
+  }
 
   getValues() { return this.form.value; }
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library-form.service.ts:245]: add return type

File: projects/admin/src/app/record/custom-editor/libraries/library.component.ts [hunk 18]
```diff
@@ -25,17 +25,20 @@ import { ToastrService } from 'ngx-toastr';
...
 })
-export class LibraryComponent implements OnInit {
+export class LibraryComponent implements OnInit, OnDestroy {
 
   public library: Library;
   public libForm: FormGroup;
-  public organisationPid;
+  public organisationPid: string;
+
+  private eventForm: Subscription;
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library.component.ts:41]: Missing doc string

File: projects/admin/src/app/record/custom-editor/libraries/library.component.ts [hunk 19]
```diff
@@ -25,17 +25,20 @@ import { ToastrService } from 'ngx-toastr';
...
 
   constructor(
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library.component.ts:43]: Missing doc string

File: projects/admin/src/app/record/custom-editor/libraries/library.component.ts [hunk 20]
```diff
@@ -55,22 +58,28 @@ export class LibraryComponent implements OnInit {
...
-        this.library = new Library({});
-        this.libForm = this.libraryForm.form;
-        this.setAsyncValidator();
-      }
+        }
+      });
     });
   }
 
+  ngOnDestroy() {
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/custom-editor/libraries/library.component.ts:79]: Missing doc string and add return type

File: projects/admin/src/app/record/detail-view/library-detail-view/library-detail-view.component.html [hunk 21]
```diff
@@ -16,15 +16,15 @@
...
     <ng-container *ngIf="record.metadata.address">
-        <dt class="col-sm-3 offset-sm-2 offset-md-0">
-            {{ 'Address' | translate }}:
-          </dt>
-          <dd class="col-sm-7 col-md-8 mb-0">
-            {{ record.metadata.address }}
-          </dd>
+      <dt class="col-sm-3 offset-sm-2 offset-md-0">
+        {{ 'Address' | translate }}:
+      </dt>
```
Comment (Garfield-fr) [inline projects/admin/src/app/record/detail-view/library-detail-view/library-detail-view.component.html:24]:
```
<dt class="col-sm-3 offset-sm-2 offset-md-0 label-title" translate>
  Address
</dt>
```

**Review by Garfield-fr — COMMENTED** [review]
With the new version of Angular, it is recommended to type the return of the function.
General: Add documentation on functions

**Review by zannkukai — APPROVED** [review]
No other remarks than @Garfield-fr

File: projects/admin/src/app/record/detail-view/library-detail-view/library-detail-view.component.html (hunk 21 above)
Comment (jma) [reply projects/admin/src/app/record/detail-view/library-detail-view/library-detail-view.component.html]: Great. I didn't now that this exists. Needs refactor other files in a separate PR.

File: projects/admin/src/app/classes/library.ts (hunk 2 above)
Comment (jma) [reply projects/admin/src/app/classes/library.ts:22]: Unfortunately this does not works with the tests.

**Pull request merged**